### PR TITLE
[Feature] Add and package the Lance JNI reader

### LINKED ISSUE #74665: [Feature] Native Lance connector for StarRocks
- [x] https://github.com/StarRocks/starrocks/pull/74659
## Is your feature request related to a problem? Please describe.

StarRocks currently supports external table connectors for Iceberg, Delta Lake, Hudi, Paimon, and other formats. However, there is no support for [Lance](https://github.com/lancedb/lance), a modern columnar data format optimized for ML workloads — embedding vectors, multi-modal data, and large-scale training datasets.

Users who store datasets in Lance format must currently ETL data into a supported format before querying with StarRocks, adding latency and operational complexity.

## Describe the solution you'd like

A native Lance connector that allows StarRocks to directly query Lance datasets on local and cloud filesystems (S3, ADLS Gen2, GCS), enabling:

- Ad-hoc exploration of ML training datasets (embeddings, document corpora)
- Quality analysis by joining Lance datasets with metadata in Iceberg/Hive tables
- Interactive analytics replacing Spark read paths on Lance tables

## Implementation Plan

The implementation is split into independently reviewable PRs:

| PR | Scope | Status |
|----|-------|--------|
| #74659 | Thrift definitions for Lance scan range fields | Open |
| #74660 | FE optimizer operators and rules | Open |
| #74661 | FE scan node and plan fragment builder | Open |
| #74662 | Java-extensions Lance JNI reader | Open |
| #74663 | BE optional pluggable connector module (follows benchmark pattern) | Open |
| #74664 | Build flag `--without-connector-lance` | Open |

**Already merged (Phase 1):** FE catalog & metadata discovery (`LanceConnector`, `LanceMetadata`, `LanceApiConverter`, `LanceTable`)

## Architecture

- **FE**: `LANCE` registered as `ConnectorType` and `TableType`. Query planning reuses `THdfsScanRange` with Lance-specific fields. CBO statistics integration.
- **BE**: Optional pluggable module following the benchmark connector pattern (`--without-connector-lance` compile-time opt-out). JNI scanner delegates to `lance-core` Java SDK.
- **java-extensions**: `LanceSplitScanner` using Arrow vectors via `lance-core:7.0.0`

## Future Work

- Predicate pushdown expression visitor
- Cloud storage credential passthrough (Azure Workload Identity, S3 IAM, GCS)
- HMS-backed catalog discovery (replacing property-based bootstrap)
- Native Rust FFI scanner (zero-copy via Arrow C Data Interface)

## Are you willing to submit PR?

Yes — PRs are already submitted (see table above).

### PR DESCRIPTION
**Pull request by beinan**
## Why I'm doing:

Lance scans require a JNI reader and its runtime dependencies in the BE distribution; building a Maven module alone does not make it deployable.

## What I'm doing:

Support Arrow LargeUtf8/LargeBinary and widen unsigned integers without overflow (UInt8/16/32/64 to SMALLINT/INT/BIGINT/DECIMAL(20,0)); align FE schema declarations with reader decoding.

Add scalar, timestamp, and list decoding and the independent reader classloader. Support Arrow date64 using UTC epoch-day conversion, including negative timestamps. Decode strings as UTF-8.

Package the scanner factory in `be/lib/jni-packages` and the reader plus runtime dependencies in `be/lib/lance-reader-lib` in this PR. Test dependencies are excluded. The optional build flag is isolated to #74664.

Use Lance 8.0.1 and matching Arrow 18.3.0 dependencies. A separate-process regression catches the Lance 7 native dispatcher class lookup failure under the BE-style isolated classloader, even when synchronous scans return rows successfully.

SQL stays disabled until #74663 adds BE descriptor dispatch and execution.

Series order: #74660 → #74661 → #74662 → #74663 → #74664. Catalog and Thrift prerequisites (#74420 and #74659) are already merged. These branches are cumulative against `main`; each preceding layer is required.

### Validation

Latest type-change validation: all 33 Lance reader tests and 3 LanceMetadataTest cases pass, including all unsigned maximum values, Unicode large strings, and binary bytes. Maven checkstyle passes.

Reader unit tests, native reads of all 7 rows across 3 fragments and multiple output chunks, and isolated-classloader regression pass. Verified both reader JAR destinations and absence of JUnit artifacts in the runtime package. Focused FE tests also pass at this PR head.

## What type of PR is this:

- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [x] Yes, this PR will result in a change in behavior.
- [ ] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [x] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [x] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
  - [ ] This pr needs auto generate documentation
- [ ] This is a backport pr

Documentation for user-visible execution is included in #74663; this series targets `main` only.

## Bugfix cherry-pick branch check:
- [ ] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 4.1
  - [ ] 4.0
  - [ ] 3.5

## Related Issue

Fixes #74665

This PR implements one layer of the linked feature. The complete implementation requires all five PRs in the series listed above.

### CODE REVIEW

**Comment by CelerData-Reviewer** [comment]
@codex review

File: java-extensions/pom.xml [hunk 1]
```diff
@@ -21,6 +21,7 @@
         <module>odps-reader</module>
         <module>iceberg-metadata-reader</module>
         <module>kudu-reader</module>
+        <module>lance-reader</module>
```
[the remark this answers is not in the document]
Comment (beinan) [reply java-extensions/pom.xml:24]: Addressed in PR #74664 which adds the build.sh packaging. See #74665 for dependency chain.

File: java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceTypeUtils.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+                return "double";
+            }
+        } else if (arrowType instanceof ArrowType.Utf8) {
+            return "string";
+        } else if (arrowType instanceof ArrowType.Binary) {
+            return "binary";
+        } else if (arrowType instanceof ArrowType.Date) {
+            return "date";
+        } else if (arrowType instanceof ArrowType.Timestamp) {
+            return "timestamp-micros";
```
[the remark this answers is not in the document]
Comment (beinan) [reply java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceTypeUtils.java]: Good point. Will address in a follow-up — current implementation handles the common case (microsecond timestamps) which covers our production Lance datasets.

**Comment by beinan** [comment]
@github-actions run-checks

File: java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceTypeUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (beinan) [reply java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceTypeUtils.java]: Addressed in 5942303: `LanceTypeUtils.fromArrowType` now dispatches on `ArrowType.Timestamp.getUnit()`. MILLISECOND maps to `timestamp-millis`; SECOND and NANOSECOND fall back to `timestamp-micros` (jni-connector `ColumnType` only recognizes millis/micros). Unit tests cover all four units.

**Comment by beinan** [comment]
@dirtysalt Conflicts have been resolved and the Lance PR stack has been rebased onto the latest main. This PR follows #74661. Could you please take another look when available? CI is rerunning on the updated commits.

File: java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceTypeUtils.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (beinan) [reply java-extensions/lance-reader/src/main/java/com/starrocks/lance/reader/LanceTypeUtils.java]: Fully addressed in 3113c53f4b. Timestamp decoding now dispatches on the Arrow unit for SECOND, MILLISECOND, MICROSECOND, and NANOSECOND, including correct floor-based handling for negative epoch values. Tests cover all four units. This supersedes the earlier partial mapping-only fix.

**Comment by beinan** [comment]
@dirtysalt Review feedback has been addressed in 3113c53f4b: recursive List/FixedSizeList handling and complete timestamp-unit decoding are implemented and tested. All review threads are resolved and the PR is mergeable. Please take another look when available.

**Comment by CelerData-Reviewer** [comment]
@codex review

**Comment by beinan** [comment]
Rebased onto the latest main (conflict-free) and fixed the Java-Extensions coverage failure.

Root cause of the `0 / 191 (0%)` report: `java-extensions/lance-reader/pom.xml` overrode surefire's `argLine` with a plain string (`--add-opens=java.base/java.nio=ALL-UNNAMED`). That discards the `-javaagent` option `jacoco:prepare-agent` injects through the `argLine` property, so the tests ran but emitted no coverage data at all. Fixed by prefixing with `@{argLine}`, matching what `udf-extensions` already does.

Note the previous BUILD failure on this PR was `Run trivy error!` — CI infrastructure, unrelated to the diff.

**Comment by beinan** [comment]
@dirtysalt — third PR in the Lance connector stack. Blocked on #74660 and #74661, but the code here is self-contained (the `lance-reader` JNI module under `java-extensions/`) so it can be reviewed independently.

Status: **no failing checks**, rebased onto latest main, no conflicts. The earlier `BUILD` failure was `Run trivy error!` — CI infrastructure, unrelated to the diff — and it cleared on re-run. The Java-Extensions coverage failure (`0 / 191`) is fixed: the module was overriding surefire's `argLine` with a plain string, which silently dropped the JaCoCo agent; it now uses `@{argLine}` like `udf-extensions`.

Your review feedback on recursive List/FixedSizeList handling and Arrow timestamp-unit decoding is implemented and covered by tests for all four units including negative epoch values.

Merge order: #74660 → #74661 → **#74662** → #74663 → #74664.
